### DISCUSSION TITLE
Flavors?

### DISCUSSION THREAD

**Discussion by nosmirck** [nuxt/nuxt#28124] [Questions] [answered]
As the title implies, how can we manage "flavors" in a Nuxt v3 app?

I know we can have different environments with different secrets and variables, but, in general that's used to configure precisely environments (development, staging, uat, production, demo, etc) but when talking about flavors there's two main uses cases:

1. white label app with core components and branded apps (that use the core functionality of the white label app and only defines styling, and minor differences in components)

2. different entry points (same app, same environments, different points of view) for instance, driver vs rider app in an uber-like application, seller vs buyer, teacher vs student, etc.

I've looked around and I haven't found any examples or documentation on how we can setup a nuxt project for both of these scenarios, I would greatly appreciate if someone can point me in the right direction, how do I setup this project? how do I add flavors? how do I add environments to each flavor? how can I define core shared components and flavor specific components? Is there any repo I can look at that contains an example?

Thanks in advance for any help!

**Comment by elisechant** [comment]
I would say that this is not a Nuxt question, this is a JavaScript application architecture question as the solution is probably quite general.

Regarding 1, Nuxt has environment variable files that can be used to support alternative configuration needed for white-labelling https://nuxt.com/docs/getting-started/configuration#environment-variables-and-private-tokens. Alternatively or in addition to, you could create JavaScript application config file(s). 

Regarding 2, you may need to create a Vue library with common components and create 2 Nuxt apps. I suggest you would set this up in a monorepo to co-locate the code.

**Reply by nosmirck** [reply]
Thank you for your answer! the environment setup I already have configured for when making one app and different environment variables depending on the environment running (prod, dev, etc). @manniL gave me the best answer by suggesting layers, it's a more broad concept than flavors but it is the most correct way to implement flavors (similar to what Android or iOS does) using Nuxt.

**Comment by TheAlexLichter** [comment, marked as answer]
For "Flavors" as in whitelabeling, Nuxt Layers is a good option. See [e.g. this video](https://www.youtube.com/watch?v=fr5yo3aVkfA)

**Reply by nosmirck** [reply]
@manniL Thank you so much for your answer, you really pointed me in the right direction. I saw the Layers in the documentation but never connected the points as of how I could use it for solving both problems since it's a very broad concept and not specific for "flavors" but it works great to solve that problem by creating a base layer and several different apps that extend the base layer, if a component from the base layer needs to be overridden, it can be copied and modified to the app's components and the app will decide to show the component in the app instead of the one in the layer. That's an amazing feature.
